### PR TITLE
Add more Mox-like functions to Req.Test

### PR DESCRIPTION
**Pull request by whatyouhide**
@wojtekmach I’m not sure how you'll feel about this but to me this is step number 1 towards what we talked about. To have complete mock functionality, we'll probably need this. It's really low maintenance (Mox has been like this for the past 7-8 years?) and if we sort of vendor its code (modified to fit this) and nimble_ownership's code you'll have top-notch testing support, IMO. Thoughts?

### CODE REVIEW

File: lib/req/test.ex [hunk 1]
```diff
@@ -167,6 +167,12 @@ defmodule Req.Test do
   @typedoc since: "0.4.15"
   @opaque stub() :: atom()
 
+  if Code.ensure_loaded?(Plug.Conn) do
+    @type plug() :: {module(), [term()]} | module() | (Plug.Conn.t() -> Plug.Conn.t())
+  else
+    @type plug() :: {module(), [term()]} | module() | (term() -> term())
```
Comment (wojtekmach) [inline lib/req/test.ex:173]: strictly speaking fun/2 is part of plug contract but fun/1 allegedly is _not_ (https://github.com/elixir-plug/plug/issues/1201#issuecomment-1872276119) so we should include it:

```suggestion
    @type plug() :: {module(), [term()]} | module() | (Plug.Conn.t() -> Plug.Conn.t() | (Plug.Conn.t(), term() -> Plug.Conn.t())
  else
    @type plug() :: {module(), [term()]} | module() | (term() -> term() | (term(), term() -> term())
```

File: lib/req/test.ex [hunk 2]
```diff
@@ -273,15 +279,21 @@ defmodule Req.Test do
...
+
   @doc """
-  Returns the stub created by `stub/2`.
+  Fetches the stubbed request handler (plug) for the given `name`.
   """
-  @spec stub(stub()) :: term()
-  def stub(stub_name) do
-    case NimbleOwnership.fetch_owner(@ownership, callers(), stub_name) do
+  @doc since: "0.5.0"
+  @spec fetch_stub!(stub()) :: plug()
```
Comment (wojtekmach) [inline lib/req/test.ex:291]: wdyt about just fetch!?

```suggestion
  @spec fetch!(stub()) :: plug()
```

File: lib/req/test.ex (hunk 1 above)
Comment (whatyouhide) [reply lib/req/test.ex:173]: Oh very good catch!!

File: lib/req/test.ex (hunk 2 above)
Comment (whatyouhide) [reply lib/req/test.ex:291]: Mh I don't think it works. I don't think `fetch_stub!` works very well either. `Req.Test.fetch!` is like... fetching what?! Maybe `Req.Test.fetch_for_name!`? I don't think anyone is expected to use this really, so I’m not sure we need a nice name, but I’m open to other ideas as well.

Comment (wojtekmach) [reply lib/req/test.ex:291]: right, let's defp it then?

Comment (whatyouhide) [reply lib/req/test.ex:291]: Yeah let's do that for now.

**Comment by wojtekmach** [comment]
Thank you!

**Pull request merged**